### ISSUE TITLE
Questions about one-shot LaNAS

### ISSUE THREAD

**Issue by yuanjunyj** [facebookresearch/LaMCTS#17]
Hi, when I try to replicate the one-shot LaNAS with the source code, I cannot get the curve as in the paper.
![image](https://user-images.githubusercontent.com/22901465/121497175-bf6e6200-ca0d-11eb-8ea1-e2700dcab6e3.png)
Then I check the code and find something strange, could you please help me work it out?
(1) In `one-shot LaNAS/LaNAS/Classifier.py`, it seems that the learning rate is too small and the learned linear model is far from the optimal. I think the learning rate should be bigger (such as 0.01), is that true?
![image](https://user-images.githubusercontent.com/22901465/121497992-95696f80-ca0e-11eb-8fbb-2d0407ac369a.png)
(2) When (1) is solved, I find that in function `search_samples_under_constraints` in `one-shot LaNAS/LaNAS/MCTS.py`, why just a pair of W and b is considered? And when I change the code into considering all Ws and bs, it seems to be hard to retrieve a sample under all constraints. How to solve this issue?
![image](https://user-images.githubusercontent.com/22901465/121497958-8a164400-ca0e-11eb-8430-f35e0923d14e.png)
(3) In the `search` function in `one-shot LaNAS/LaNAS/MCTS.py`, the tree is updated after each sample is evaluated, but in other scripts (like Distributed_LanAS and LaNAS_NASBench101), the tree is updated after evaluating 20/50 samples. How often should the "learning phase" be recondutced (or in other words, how many samples should be retrieved in each "searching phase")?
(4) Which Cp should be chosen to replicate the result of one-shot LaNAS? I find it very big (Cp=10) in the code and it seems a little bit unreasonable.

**Comment by linnanwang** [comment]
Hello Jun,

Thanks for your interests, 

1) you should try out to see the loss progress: you can start with a small lr and longer epochs.

2) W, b are array of constraints, you need take a look at the following codes:
[image](https://user-images.githubusercontent.com/4675611/121544608-b73b0600-c9be-11eb-8b78-4a728d6d73bd.png)

3) You can update the tree when you receive a new sample to keep the tree up to date.

4) If the accuracy is in the range of [0, 100], then it is 10; if the range is [0, 1], it is 0.1

Hope that helps.

**Comment by linnanwang** [comment]
I admit this version is not well implemented, but here might be a better version if you want to play with it:

https://github.com/facebookresearch/LaMCTS/tree/master/LA-MCTS

**Comment by yuanjunyj** [comment]
Thank you so much. It is great to have another implemented version of LA-MCTS!
My target is to reproduce the oneshot LaNAS results and try to make improvements aseed on your code, and still I have some questions.
(1)  I understand the meaning of W and b and how it works, but in the sampling process, `rand_arch` is returned once it meet one of the constraints according to your code.
(2) In one-shot LaNAS, `init_train` in MCTS is dismissed (while it appears in other two MCTS implementations of LaNAS). Should it be implememented? And how should it be implemented?
(3) Do the hyper-params keep the same among all experiments as mentioned in the paper? I find the tree-height is set to 5 by default in the code.
![image](https://user-images.githubusercontent.com/22901465/121560564-3aee0480-ca4a-11eb-955e-764ddc4cf617.png)
It is very kind of you if you can provide the code or more details about reproducing the result of one-shot LaNAS

**Comment by linnanwang** [comment]
While you're claiming not producing the results, can you please first let me know the results on your side? Please note you are expected to run multiple times to reproduce the figure.

After having the initial results, let's talk about the difference then.

**Comment by yuanjunyj** [comment]
After I fixed the issue of learning rate, I have run the algorithm for 5 trials.
Currently, I set the tree height to 5 and used 200 samples for initialization. The tree will be updated when receiving a new sample, that is, #select=1. The current test_acc - UVs curve is attached below. Different color indicates a different trial.
![image](https://user-images.githubusercontent.com/22901465/121845746-79204800-cd18-11eb-953f-de968a3bf3e4.png)
Two of them have evaluated over 2000 samples and obtained the best test accuracy on supernet of 81.69%(found in sample#2306) and 81.45%(found in sample#1524)
Other three trials are still running. In the latest results they obtained the best test accuracy on supernet of 81.47%(found in sample#38), 80.29%(found in sample#779), and 81.59%(found in sample#705), separately.
I think some of the params I used is incorrect. Could you please help me find out the correct combination of params used in one-shot LaNAS?

**Comment by linnanwang** [comment]
Can you add legend to your figure?

**Comment by yuanjunyj** [comment]
![image](https://user-images.githubusercontent.com/22901465/121906545-cbd02300-cd5d-11eb-981c-57aa3686fb9c.png)
Lines of different color just indicate different trials. The figure is a little different from the above one just because I collect the results of more samples. 
Besides, I have sent you an email in which I described my problems more detailedly:)
Thanks for your patiency!

**Comment by linnanwang** [comment]
Thanks, can you please send the email to wangnan318@gmail.com? These results seem interesting, let me also check back my log here.

**Comment by yuanjunyj** [comment]
I have sent the email.
If you did not receive it, please let me know :)

**Comment by linnanwang** [comment]
Still I did not recieve the email, but do you notice that even random search can reach 0.83 on the supernet?

**Comment by linnanwang** [comment]
Closed for solving the problem via email.

![image](https://user-images.githubusercontent.com/4675611/122577583-d0b7ff80-d007-11eb-9189-c4eae9eb227e.png)
